proof/crefine/RISCV64/Recycle_C.thy

```diff
@@ -805,16 +805,123 @@ lemma ntfn_q_refs'_no_NTFNBound[simp]:
+                    and val="from_bool (scSporadic sc)"
+                    and R="obj_at' (\<lambda>tcb. tcbSchedContext tcb = Some scPtr) t and ko_at' sc scPtr
+                           and no_0_obj'"
+                    and R'=UNIV
+                     in ccorres_symb_exec_r_known_rv)
+           apply (rule conseqPre, vcg)
+           apply normalise_obj_at'
+           apply (frule (1) obj_at_cslift_tcb)
+           apply (frule (1) obj_at_cslift_sc)
+           apply (clarsimp simp: typ_heap_simps ctcb_relation_def csched_context_relation_def
+                                 to_bool_def
+                          split: if_splits)
+          apply ceqv
+         apply (rule ccorres_split_nothrow[where r'=dc and xf'=xfdc])
+             apply (rule ccorres_cond[where R=\<top>])
+               apply fastforce
+              apply (rule ccorres_move_c_guard_tcb)
+              apply (rule_tac Q="\<lambda>s. cur_sc = ksCurSc s
+                                     \<and> obj_at' (\<lambda>tcb. tcbSchedContext tcb = Some scPtr) t s"
+                           in ccorres_cond_both'[where Q'=\<top>])
+                apply clarsimp
+                apply (frule (1) obj_at_cslift_tcb)
+                apply (fastforce dest!: rf_sr_ksCurSC simp: typ_heap_simps' ctcb_relation_def)
+               apply (rule ccorres_move_c_guard_tcb)
+               apply (ctac add: refill_unblock_check_ccorres)
+              apply (rule ccorres_return_Skip)
+             apply (rule ccorres_return_Skip)
+            apply ceqv
+           apply (ctac add: possibleSwitchTo_ccorres)
+          apply wpsimp
+         apply (vcg exspec=refill_unblock_check_modifies)
+        apply (vcg exspec=sc_sporadic_modifies)
+       apply (rule_tac xf'="ret__unsigned_long_'"
+                   and val="from_bool False"
+                   and R="obj_at' (\<lambda>tcb. tcbSchedContext tcb = None) t"
+                   and R'=UNIV
+                    in ccorres_symb_exec_r_known_rv)
+          apply (rule conseqPre, vcg)
+          apply clarsimp
+          apply (frule (1) obj_at_cslift_tcb)
+          apply (clarsimp simp: typ_heap_simps ctcb_relation_def)
+         apply ceqv
+        apply clarsimp
+        apply (ctac add: possibleSwitchTo_ccorres)
+       apply (vcg exspec=sc_sporadic_modifies)
+      apply (rule_tac Q'="\<lambda>_. tcb_at' t and valid_objs' and no_0_obj'
+                              and pspace_aligned' and pspace_distinct' and pspace_bounded'
+                              and (\<lambda>s. ksCurDomain s \<le> maxDomain)
+                              and (\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s)"
+                   in hoare_post_imp)
+       apply (fastforce simp: obj_at'_def)
+      apply wpsimp
+     apply (vcg exspec=setThreadState_modifies)
+    apply (ctac add: setThreadState_ccorres)
+   apply (fastforce simp: mask_def[where n=4] ThreadState_defs typ_heap_simps ctcb_relation_def
```

Comment:
don't know what this 4 is; should we be concerned about this magic number?

Reply:
I stole this from somewhere else. The `4` comes from the guard/precondition of `setThreadState_ccorres`:

```
(⦃∀cl fl. cthread_state_relation_lifted st (cl⦇tsType_CL := ´ts && mask 4⦈, fl)⦄
      ∩ ⦃´tptr = tcb_ptr_to_ctcb_ptr thread⦄)
```

There's unfortunately no comment there but I think it stems from the body of `thread_state_ptr_set_tsType`. It's pretty gory that low down. 

I am actually able to remove the `n=4`. Would it be better to just have it as `mask_def`, or is it generally unwise to unfold `mask_def`? I regrettably have very little experience working with masking and bit operations like this, so I'm not sure of the rules and recommendations.

Reply:
If there is no other `mask` in the goal, it's safe to unfold. However: somewhere in AArch64 I made a `simp` rule that gets rid of all of these automatically for thread states. We should probably increase the use of that one instead. It's [here](https://github.com/seL4/l4v/blob/53f44dea1e7763b94a98037629a3c608a9a06c44/proof/crefine/AARCH64/Wellformed_C.thy#L668-L671), it should copy over fine to MCS and other architectures.

Reply:
I did copy that lemma across to RISCV MCS, but I was wanting the version for `Inactive`, so I added that one, too. Did you want me to just have these for RISCV? Or should I put these into the other arches for MCS?

Reply:
I think unfolding `mask_def` in general is not that great, but sometimes it's unavoidable, and as this is a terminal step (solves the subgoal), I have no idea, but it's not too bad. The `n=4` though is bizarre to read, and since you said you can remove it, thumbs up to doing that.

Regarding where to put it... that is a good question. Ultimately this will end up in all arches, but I'm a bit lost on how we're managing MCS given the one-arch push to finish CRefine. My suggestion: leave it on RISCV only for MCS, but backport to master in another PR, and then all the arches will pick that up in the next merge.